torch_geometric/utils/convert.py

```diff
@@ -614,3 +614,178 @@ def from_dgl(
+
+    Args:
+        G (networkx.Graph or networkx.DiGraph): A networkx graph.
+        node_type_attribute (str): The attribute containing the type of a
+            node. For the resulting structure to be valid, this attribute
+            must be set for every node in the graph.
+        edge_type_attribute (str): The attribute containing the type of
+            an edge. If set to :obj:`None`, the value :obj:`"to"` will be
+            used in the final structure. Otherwise, this attribute must be
+            set for every edge in the graph. (default: :obj:`None`)
+
+    Example:
+
+        >>> data = from_hetero_networkx(G, node_type_attribute="type",
+        ...                    edge_type_attribute="type")
+        <torch_geometric.data.HeteroData()>
+
+    :rtype: :class:`torch_geometric.data.HeteroData`
+    """
+    import networkx as nx
+
+    from torch_geometric.data import HeteroData
+
+    def get_edge_attributes(G, edge_indexes: list,
+                            edge_attrs: list = None) -> dict:
+        r"""Collects the attributes of a list of graph edges in a dictionary.
+
+        Args:
+            G (networkx.Graph or networkx.DiGraph): A networkx graph.
+            edge_indexes (list, optional): The list of edge indexes whose
+                attributes are to be collected. If set to :obj:`None`, all
+                edges of the graph will be included. (default: :obj:`None`)
+            edge_attrs (list, optional): The list of expected attributes to
+                be found in every edge. If set to :obj:`None`, the first
+                edge encountered will set the values for the rest of the
+                process. (default: :obj:`None`)
+
+        Raises:
+            ValueError: If some of the edges do not share the same list
+            of attributes as the rest, an error will be raised.
+        """
+        data = defaultdict(list)
+        edge_to_data = list(G.edges(data=True))
+
+        for edge_index in edge_indexes:
+            _, _, feat_dict = edge_to_data[edge_index]
+            if edge_attrs is None:
+                edge_attrs = feat_dict.keys()
+            if set(feat_dict.keys()) != set(edge_attrs):
+                raise ValueError('Not all edges contain the same attributes.')
+            for key, value in feat_dict.items():
+                data[str(key)].append(value)
+
+        return data
+
+    def get_node_attributes(G, nodes: list,
+                            expected_node_attrs: list = None) -> dict:
+        r"""Collects the attributes of a list of graph nodes in a dictionary.
+
+        Args:
+            G (networkx.Graph or networkx.DiGraph): A networkx graph.
+            nodes (list, optional): The list of nodes whose attributes are to
+                be collected. If set to :obj:`None`, all nodes of the graph
+                will be included. (default: :obj:`None`)
+            node_attrs (list, optional): The list of expected attributes to
```

Comment:
Lets expose this argument in `from_hetero_netrworkx`

Reply:
Should I expose `nodes ` or `node_attrs`?

Reply:
I went for the `nodes` in my last commit.

Reply:
`node_attrs`, so that the user can specify which attributes a node has, similar to `from_networkx`.

Reply:
Sure! But since HeteroData has different groups of nodes, what implementation do you think would be the best?

1. `node_attrs` is a dictionary {node_group --> attributes} where each type of node has its own list of attributes to stack.

```python
node_attrs = {
type_1: ['y', 'z'],
type_2: ['u', 'v'],
type_3: ['u', 'y']
}
``` 

2. `node_attrs` is a single set of attributes. For each group of nodes, we expect all of them to be present.

```python
node_attrs = {'u', 'v', 'w'} # All required for every node group
``` 

3.  `node_attrs` is a single set of attributes. For each group of nodes, we only stack the ones that are available without raising an error if some are missing.

```python
node_attrs = {'u', 'v', 'w'} # To get when possible for every node group
``` 

n°3 is semi-flexible: we can allow some groups of nodes to have different features on the condition that no two groups of different nodes have the same attribute names for different purposes (assuming we want one but not the other).

I think n°1 is probably the best since we can express any situation, but it is also more complex to use for the user.

What do you think?

Reply:
In my latest commit, I implemented `attrs` to accommodate both node and edge attributes. 
It is a dictionary that maps one group to the attributes to concatenate. It looks like: 

```python
attrs = {
    "A": ["u", "v", "c"],
    "B": ["w", "z", "c"],
    ("A", "to", "A"): ["a"],
    ("B", "to", "B"): ["b"]
}
```

Reply:
Sorry for the later reply, but lets keep it simple and go for solution 2 because
1. It keeps the API similar to [to_netowrkx](https://pytorch-geometric.readthedocs.io/en/latest/modules/utils.html#torch_geometric.utils.to_networkx) where `node_attrs` and `edge_attrs` are `List[str]`, and all nodes need to have those attributes. 
2. I like the added flexibility of 1, but we should only add more flexiblity (and more complexity) if the end user really needs it. I don't think that most users need this flexibility. 

I see that you have already made some changes. Apologies for asking to change it again.

Reply:
No worries, I knew that it was subject to change.
I'll make the update for solution n°2 then, thanks!